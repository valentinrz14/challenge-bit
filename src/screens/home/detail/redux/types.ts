import { HomeListData } from '@api/types';

/*
 ** Types
 */
export interface HomeDetailState {
  loading: boolean;
  operationDetail: HomeListData | null;
  error: string | null | unknown;
}

export interface HomeDetailActionTypes {
  type: string;
  payload: HomeDetailState;
}
