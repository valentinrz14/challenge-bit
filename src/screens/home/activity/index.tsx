import { HomeListData } from '@api/types';
import { LineHorizontal } from '@core/components/line-horizontal';
import { H4 } from '@core/components/Typography';
import { Colors } from '@core/styles/colors';
import { RenderItemList } from '@screens/home/home-list/render-item-list';
import { ReduxRootState } from '@store/types';
import React, { FunctionComponent } from 'react';
import { FlatList, View } from 'react-native';
import { useSelector } from 'react-redux';
import { styles } from './activity-styles';

/*
 ** Types
 */

interface ActivityScreenProps {}

/*
 ** Component
 */

export const ActivityScreen: FunctionComponent<ActivityScreenProps> = () => {
  const { operationList, loading } = useSelector(
    (state: ReduxRootState) => state.home,
  );

  const renderItem = ({ item }: { item: HomeListData }) => {
    return <RenderItemList item={item} />;
  };

  const itemSeparatorComponent = () => <View style={styles.itemSeparator} />;

  const keyExtractor = (item: HomeListData) => item.id;

  return (
    <View style={styles.mainContainer}>
      <LineHorizontal height={1} bgColor={Colors.blue} />
      <H4 color="black" fontFamily="MontserratMedium" textStyle={styles.title}>
        Todos los Movimientos
      </H4>
      <FlatList
        style={styles.container}
        contentContainerStyle={styles.contentList}
        data={operationList}
        keyExtractor={keyExtractor}
        renderItem={renderItem}
        ItemSeparatorComponent={itemSeparatorComponent}
        refreshing={loading}
      />
    </View>
  );
};
