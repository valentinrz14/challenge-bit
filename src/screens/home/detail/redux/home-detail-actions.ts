import { HomeListData } from '@api/types';
import {
  HOME_DETAIL_ADD_OPERATION,
  HOME_DETAIL_SET_LOADING,
  HOME_DETAIL_SET_ERROR,
} from '@screens/home/detail/redux/names';

export const homeDetailSetLoading = () => ({
  type: HOME_DETAIL_SET_LOADING,
  payload: {
    loading: true,
  },
});

export const homeDetailAddOperationDetail = (data: HomeListData) => ({
  type: HOME_DETAIL_ADD_OPERATION,
  payload: {
    loading: false,
    operationDetail: data,
  },
});

export const homeDetailSetError = (error: string) => ({
  type: HOME_DETAIL_SET_ERROR,
  payload: {
    loading: false,
    error,
  },
});
