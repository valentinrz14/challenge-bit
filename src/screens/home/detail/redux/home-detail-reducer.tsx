import {
  HOME_DETAIL_ADD_OPERATION,
  HOME_DETAIL_SET_ERROR,
  HOME_DETAIL_SET_LOADING,
} from '@screens/home/detail/redux/names';
import {
  HomeDetailActionTypes,
  HomeDetailState,
} from '@screens/home/detail/redux/types';

/*
 ** Constants
 */

const initialState: HomeDetailState = {
  loading: false,
  operationDetail: null,
  error: null,
};

export const HomeDetailReducer = (
  state = initialState,
  action: HomeDetailActionTypes,
) => {
  if (action.type === HOME_DETAIL_SET_LOADING) {
    return {
      ...state,
      loading: true,
    };
  }

  if (action.type === HOME_DETAIL_ADD_OPERATION) {
    return {
      ...state,
      operationDetail: action.payload.operationDetail,
      loading: false,
    };
  }

  if (action.type === HOME_DETAIL_SET_ERROR) {
    return {
      ...state,
      error: action.payload.error,
      loading: false,
    };
  }

  return state;
};
