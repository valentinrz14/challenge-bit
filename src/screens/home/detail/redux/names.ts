export const HOME_DETAIL_SET_LOADING = '@HOME_DETAIL/SET_LOADING';
export const HOME_DETAIL_ADD_OPERATION = '@HOME_DETAIL/ADD_OPERATION';
export const HOME_DETAIL_SET_ERROR = '@HOME_DETAIL/SET_ERROR';
