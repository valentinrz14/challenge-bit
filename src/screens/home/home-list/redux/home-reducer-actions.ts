import { HomeListData } from '@api/types';
import {
  HOME_ADD_OPERATION_LIST,
  HOME_SET_LOADING,
  HOME_SET_ERROR,
} from '@screens/home/home-list/redux/names';

export const homeAddOperationList = (data: HomeListData) => {
  return {
    type: HOME_ADD_OPERATION_LIST,
    payload: {
      operationList: data,
      loading: false,
    },
  };
};

export const homeSetLoading = () => ({
  type: HOME_SET_LOADING,
  payload: {
    loading: true,
  },
});

export const homeSetError = (error: string) => ({
  type: HOME_SET_ERROR,
  payload: { error, loading: false },
});
