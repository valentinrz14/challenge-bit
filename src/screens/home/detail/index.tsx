import React, { FunctionComponent } from 'react';
import { Text, View } from 'react-native';
import { Colors } from '@core/styles/colors';
import { RenderContentDetail } from '@screens/home/detail/render-content-detail';
import { RenderHeaderDetail } from '@screens/home/detail/render-header-detail';
import { LineHorizontal } from '@core/components/line-horizontal';
import { HomeStackParamsList } from '@navigation/home/types';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { styles } from '@screens/home/detail/detail-styles';
import { useSelector } from 'react-redux';
import { ReduxRootState } from '@store/types';
import { LayoutLoading } from '@core/components/layout-loading';

/*
 ** Types
 */
type DetailsScreenNavigationProp = NativeStackScreenProps<
  HomeStackParamsList,
  'Details'
>;

interface DetailsScreenProps extends DetailsScreenNavigationProp {}

/*
 ** Component
 */
export const DetailsScreen: FunctionComponent<DetailsScreenProps> = () => {
  const { operationDetail, loading } = useSelector(
    (state: ReduxRootState) => state.homeDetail,
  );

  if (loading) {
    return <LayoutLoading size="large" />;
  }

  const backgroundColor =
    operationDetail?.state === 'Rechazado' ? Colors.red : Colors.green;

  return (
    <>
      <LineHorizontal height={1} bgColor={backgroundColor} />
      <View style={styles.mainContainer}>
        <RenderHeaderDetail state={operationDetail?.state} />
        <View style={styles.container}>
          <Text>Detalle de operacion</Text>
          <View style={styles.content}>
            <RenderContentDetail
              title="Operacion"
              subtitle={operationDetail?.title}
            />
            <RenderContentDetail title="ID" subtitle={operationDetail?.id} />
            <RenderContentDetail
              title="Fecha"
              subtitle={operationDetail?.date}
            />
            <RenderContentDetail
              title="Importe"
              subtitle={`$${operationDetail?.currencyAmount} ${operationDetail?.currencySymbol}`}
            />
            <RenderContentDetail
              title="Destino"
              subtitle={operationDetail?.destination}
            />
          </View>
        </View>
      </View>
    </>
  );
};
