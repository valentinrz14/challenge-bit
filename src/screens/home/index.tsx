import { HomeStackParamsList } from '@navigation/home/types';
import { DashboardStackParamsList } from '@navigation/types';
import { BottomTabScreenProps } from '@react-navigation/bottom-tabs';
import { CompositeScreenProps } from '@react-navigation/native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { styles } from '@screens/home/home-styles';
import React, { FunctionComponent } from 'react';
import { SafeAreaView, FlatList, View, Text } from 'react-native';
import { useSelector } from 'react-redux';
import { ReduxRootState } from '@store/types';
import { HomeListData } from '../../api/types';
import { RenderHeaderList } from './home-list/render-header-list';
import { RenderItemList } from './home-list/render-item-list';
/*
 ** Types
 */
type HomeScreenNavigationProp = CompositeScreenProps<
  BottomTabScreenProps<DashboardStackParamsList, 'HomeStack'>,
  NativeStackScreenProps<HomeStackParamsList>
>;

interface HomeScreenProps extends HomeScreenNavigationProp {}

/*
 ** Component
 */
const keyExtractor = (item: HomeListData) => item.id;

const ListEmptyComponent = () => (
  <View style={styles.containerListEmpty}>
    <Text style={styles.titleListEmpty}>No hay movimientos</Text>
  </View>
);

const renderItem = ({ item }: { item: HomeListData }) => {
  return <RenderItemList item={item} />;
};

const itemSeparatorComponent = () => <View style={styles.itemSeparator} />;

export const HomeScreen: FunctionComponent<HomeScreenProps> = () => {
  const { shortOperationList, loading } = useSelector(
    (state: ReduxRootState) => state.home,
  );

  return (
    <SafeAreaView style={styles.mainContainer} testID="home-screen">
      <RenderHeaderList
        testID="home-screen-header"
        showMoreLastMovements={shortOperationList.length >= 5}
      />
      <FlatList
        testID="home-screen-list"
        style={styles.container}
        data={shortOperationList}
        ListEmptyComponent={ListEmptyComponent}
        refreshing={loading}
        keyExtractor={keyExtractor}
        contentContainerStyle={styles.contentList}
        renderItem={renderItem}
        ItemSeparatorComponent={itemSeparatorComponent}
      />
    </SafeAreaView>
  );
};
