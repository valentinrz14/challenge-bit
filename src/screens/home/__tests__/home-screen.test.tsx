import React from 'react';
import { render } from '@testing-library/react-native';
import { HomeScreen } from '@screens/home';
import { Provider } from 'react-redux';
import { Store, AnyAction } from 'redux';
import { createTestStore } from '../../../../__tests__/utils/create-test-store';

jest.useFakeTimers();

jest.mock('react-query', () => ({
  useQuery: () => ({ isLoading: false, error: {}, data: {} }),
}));

let store: Store<any, AnyAction>;
const routeMock: any = { params: undefined, screen: 'Home' };
const navigationMock: any = {
  navigate: () => {},
};

beforeEach(() => {
  store = createTestStore();
});

describe('Home Screen', () => {
  it('should render screen with one operation', () => {
    const { getByTestId, toJSON } = render(
      <Provider store={store}>
        <HomeScreen route={routeMock} navigation={navigationMock} />
      </Provider>,
    );

    expect(getByTestId('home-screen')).toBeTruthy();
    expect(getByTestId('home-screen-header')).toBeTruthy();
    expect(getByTestId('home-screen-list')).toBeTruthy();

    expect(toJSON).toMatchSnapshot();
  });
});
