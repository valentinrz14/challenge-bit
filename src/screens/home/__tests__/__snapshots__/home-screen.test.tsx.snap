// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should render screen with one operation 1`] = `[Function]`;
